### PR TITLE
Port tests from setContext.ts over to Rust

### PR DESCRIPTION
**Pull request by briannafugate408**
This pull request introduces several improvements and bug fixes to the handling and validation of the `@fromContext` directive in Apollo Federation. The changes focus on stricter validation, clearer error messages, improved test coverage, and a minor bug fix in argument validation. The most significant updates are grouped below.

Changes included: 
* Added validation to prevent `@fromContext` from being used on directive definition arguments, returning a specific error if encountered.
* Improved error messages for `@fromContext` misuse, including more precise context in messages and better handling of missing or malformed context references.
* Updated the validator logic to continue on missing context/selection instead of panicking, ensuring all errors are collected.
* Fixed the context parsing logic to return the unmatched input as the selection when no valid selection is present, and updated related tests for edge cases.
* Port tests from https://github.com/apollographql/federation/blob/main/composition-js/src/__tests__/compose.setContext.test.ts over to Rust implementation and implement transformation for spec directives to make context tests work. The transform logic on arguments can be found here:https://github.com/apollographql/federation/blob/main/internals-js/src/specs/contextSpec.ts#L47-L53

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] PR description explains the motivation for the change and relevant context for reviewing
- [x] PR description links appropriate GitHub/Jira tickets (creating when necessary)

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: A lot of (if not most) features benefit from built-in observability and `debug`-level logs. Please read [this guidance](https://github.com/apollographql/router/blob/dev/dev-docs/metrics.md#adding-new-metrics) on metrics best-practices.
[^4]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[FED-688]: https://apollographql.atlassian.net/browse/FED-688?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-federation/src/link/context_spec_definition.rs [hunk 1]
```diff
@@ -52,6 +57,25 @@ impl ContextSpecDefinition {
...
     }
 
+    fn static_argument_transform(
+        subgraph: &Subgraph<Validated>,
+        args: IndexMap<Name, Value>,
+    ) -> IndexMap<Name, Value> {
+        let subgraph_name = &subgraph.name;
+        args.into_iter()
+            .map(|(key, value)| {
+                if key.as_str() == "name" {
```
Comment (tninesling) [inline apollo-federation/src/link/context_spec_definition.rs]: nit: We usually use the constant `FEDERATION_NAME_ARGUMENT_NAME` from the Federation spec file for this

File: apollo-federation/src/link/context_spec_definition.rs [hunk 2]
```diff
@@ -147,7 +171,7 @@ impl SpecDefinition for ContextSpecDefinition {
             ],
             true,
             Some(&|v| CONTEXT_VERSIONS.get_dyn_minimum_required_version(v)),
-            None, // TODO: Add transform
+            Some(Rc::new(Self::static_argument_transform)),
```
Comment (tninesling) [inline apollo-federation/src/link/context_spec_definition.rs:174]: 🚀 Nice!

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 3]
```diff
@@ -0,0 +1,1494 @@
...
+        }
+
+        type U @key(fields: "id") {
+          id: ID!
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_ok(), "Expected composition to succeed");
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: This assertion is wrong. This case should reject the use of context with the error `'[Subgraph1] @fromContext argument cannot be used on a directive definition "@foo(a:)".'` per [source](https://github.com/apollographql/federation/blob/main/composition-js/src/__tests__/compose.setContext.test.ts#L1236C1-L1237C1).

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 4]
```diff
@@ -0,0 +1,1494 @@
...
+    let supergraph = result.expect("Expected composition to succeed");
+    assert!(
+        !supergraph.schema().schema().types.is_empty(),
+        "Supergraph should contain types"
+    );
+
+    // Note: The JavaScript test checks for hints, but the Rust implementation
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: The key assertion is actually that we produce the right hint in this case. Can you update this to check that we output the correct [code](https://github.com/apollographql/federation/blob/main/composition-js/src/__tests__/compose.setContext.test.ts#L1177) and [message](https://github.com/apollographql/federation/blob/main/composition-js/src/__tests__/compose.setContext.test.ts#L1185)?

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 5]
```diff
@@ -0,0 +1,1494 @@
...
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains("Argument \"U.field(a:)\" is contextual in at least one subgraph but in \"U.field\" it does not have @fromContext, is not nullable and has no default value."),
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: This should be a strict equality check. If we're accidentally adding anything to the beginning or end of this message, we should catch that here.

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 6]
```diff
@@ -0,0 +1,1494 @@
...
+          id: ID!
+          field: Int! @shareable
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_ok(), "Expected composition to succeed");
+    let supergraph = result.expect("Expected composition to succeed");
+    assert!(
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: nit: It looks like this was omitted from the only test that actually had it, but it leaked into all these other tests. I don't think we're running the risk of generating empty schemas, since they wouldn't validate after merge. Feel free to remove these checks.

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 7]
```diff
@@ -0,0 +1,1494 @@
...
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    println!("error_message: {}", error_message);
+    assert!(
+        error_message.contains("[Subgraph1] Context \"context\" is used in \"U.field(a:)\" but the selection is invalid: One of the types in the selection is an interfaceObject: \"T\"."),
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: This should also be an exact string match

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 8]
```diff
@@ -0,0 +1,1494 @@
...
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains("[Subgraph1] Context \"context\" is used in \"U.field(a:)\" but the selection is invalid: directives are not allowed in the selection"),
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: This should also be an exact string match

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 9]
```diff
@@ -0,0 +1,1494 @@
...
+    assert!(
+        error_message.contains("[Subgraph1] Context \"context\" is used in \"U.field(a:)\" but the selection is invalid: multiple selections are made"),
+        "Expected error message about invalid context selection with multiple fields, but got: {}",
+        error_message
+    );
+}
+
+#[test]
+#[ignore] // Skipped test - nullability mismatch validation
+fn nullability_mismatch_not_ok_if_argument_non_nullable() {
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: The JS source says this test is invalid now, so I think we can just delete it

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 10]
```diff
@@ -0,0 +1,1494 @@
...
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    println!("error_message: {}", error_message);
+    assert!(
+        error_message.contains("[Subgraph1] Context \"context\" is used in \"U.field(a:)\" but the selection is invalid for type \"T2\"."),
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: The [JS source's](https://github.com/apollographql/federation/blob/main/composition-js/src/__tests__/compose.setContext.test.ts#L602) error message ends with `Error: Cannot query field "prop" on type "T".`. Would be good to make sure that's in the assertion.

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 11]
```diff
@@ -0,0 +1,1494 @@
...
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains("Context and selection must be present"),
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: It looks like [this message](https://github.com/apollographql/federation/blob/main/composition-js/src/__tests__/compose.setContext.test.ts#L407) should be `'[Subgraph1] @fromContext argument does not reference a context "{ prop }".'`

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 12]
```diff
@@ -0,0 +1,1494 @@
...
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains("[Subgraph1] Context \"context\" is used in \"U.field(a:)\" but the selection is invalid for type \"T\"."),
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs]: Are we generating the `Error: Cannot query field "invalidprop" on type "T".` at the end?

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 13]
```diff
@@ -0,0 +1,1494 @@
...
+
+        type U @key(fields: "id") {
+          id: ID!
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_ok(), "Expected composition to succeed");
+    let supergraph = result.expect("Expected composition to succeed");
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs:42]: Wouldn't it be more concise to delete line 42 since 43 achieves the same thing?

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 14]
```diff
@@ -0,0 +1,1494 @@
...
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: Same suggestion as lines 42 & 43

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 15]
```diff
@@ -0,0 +1,1494 @@
...
+        }
+
+        type U @key(fields: "id") {
+          id: ID!
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: Same applies to all similar cases.

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 16]
```diff
@@ -0,0 +1,1494 @@
...
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains(
+            "[Subgraph1] Context name \"\" is invalid. It should have only alphanumeric characters."
+        ),
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: Changing this to use `assert_eq!(error_message, "....")`  would be closer to JS's `expect(...).toBe(...)`

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 17]
```diff
@@ -0,0 +1,1494 @@
...
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains(
+            "[Subgraph1] @fromContext arguments may not have a default value: \"U.field(a:)\"."
+        ),
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: Switch to `assert_eq!()`

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 18]
```diff
@@ -0,0 +1,1494 @@
...
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert!(
+        error_message.contains("[Subgraph1] @fromContext argument cannot be used on a field implementing an interface field \"U.field(a:)\"."),
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: This string doesn't match the [JS string](https://github.com/apollographql/federation/blob/f3ab499eaf62b1a1c0f08b838d2cbde5accb303a/composition-js/src/__tests__/compose.setContext.test.ts#L1331)

File: apollo-federation/tests/composition/compose_set_context.rs (hunk 4 above)
Comment (conwuegb) [reply apollo-federation/tests/composition/compose_set_context.rs]: The helper function, [assert_hints_equal()](https://github.com/apollographql/router/blob/494889cc743dcd89566ddca5964e238b5d9600e4/apollo-federation/tests/composition/mod.rs#L145) does this for `CompositionHint`s

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 19]
```diff
@@ -0,0 +1,1494 @@
...
+
+        type U @key(fields: "id") {
+          id: ID!
+          field(a: String): Int! @shareable
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_ok(), "Expected composition to succeed");
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs:375]: Redundant

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 20]
```diff
@@ -0,0 +1,1494 @@
...
+
+        type U @key(fields: "id") {
+          id: ID!
+          field(a: String!): Int! @shareable
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_err(), "Expected composition to fail");
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: Redundant

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 21]
```diff
@@ -0,0 +1,1494 @@
...
+        }
+
+        type U @key(fields: "id") {
+          id: ID!
+        }
+        "#,
+    };
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+    assert!(result.is_ok(), "Expected composition to succeed");
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: Since you probably have deleted this, my comment is invalid, but in case it does stick around -- this is [supposed to fail](https://github.com/apollographql/federation/blob/f3ab499eaf62b1a1c0f08b838d2cbde5accb303a/composition-js/src/__tests__/compose.setContext.test.ts#L682-L687).

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 22]
```diff
@@ -0,0 +1,1494 @@
...
+    let supergraph = result.expect("Expected composition to succeed");
+    assert!(
+        !supergraph.schema().schema().types.is_empty(),
+        "Supergraph should contain types"
+    );
+}
+
+#[test]
+fn vanilla_setcontext_success_case_alt() {
```
Comment (conwuegb) [inline apollo-federation/tests/composition/compose_set_context.rs]: This was commented out in the JS code -- was it supposed to be included here?

**Review by conwuegb — CHANGES_REQUESTED** [review]

File: apollo-federation/tests/composition/compose_set_context.rs (hunk 12 above)
Comment (briannafugate408) [reply apollo-federation/tests/composition/compose_set_context.rs]: The only error being generated is:

`[SubgraphError { subgraph: "Subgraph1", error: ContextSelectionInvalid { message: "Context \"context\" is used in \"U.field(a:)\" but the selection is invalid for type \"T\"." }, locations: [] }]`

File: apollo-federation/tests/composition/compose_set_context.rs (hunk 11 above)
Comment (briannafugate408) [reply apollo-federation/tests/composition/compose_set_context.rs]: Need to dig into this one because the only message we get is: 

 `error_message: [Subgraph1] An internal error has occurred, please report this bug to Apollo.`

File: apollo-federation/tests/composition/compose_set_context.rs (hunk 10 above)
Comment (briannafugate408) [reply apollo-federation/tests/composition/compose_set_context.rs]: It's not per the `println!`. For those assertions, should we be updating the error messages?

File: apollo-federation/src/schema/validators/from_context.rs [hunk 23]
```diff
@@ -1029,12 +1033,17 @@ mod tests {
...
         // We expect an error for the @fromContext on a field implementing an interface
+        assert!(
+            !errors.errors.is_empty(),
+            "Expected errors for @fromContext on interface implementation, but got none"
+        );
         assert!(
             errors.errors.iter().any(|e| matches!(
                 e,
-                SingleFederationError::ContextNotSet { message } if message == "@fromContext argument cannot be used on a field implementing an interface field \"User.id(contextArg:)\"."
+                SingleFederationError::ContextNotSet { message } if message == "@fromContext argument cannot be used on a field implementing an interface field \"Entity.id\"."
```
Comment (tninesling) [inline apollo-federation/src/schema/validators/from_context.rs:1043]: This change looks wrong. Isn't the `@fromContext` on `User.id(contextArg:)`?

File: apollo-federation/src/schema/validators/from_context.rs [hunk 24]
```diff
@@ -1075,7 +1084,33 @@ mod tests {
             &context_map,
             &mut errors,
         )
-        .expect_err("validates fromContext directives");
+        .expect("validates fromContext directives");
+
+        // Should have errors for invalid context reference and missing selection
+        assert!(
```
Comment (tninesling) [inline apollo-federation/src/schema/validators/from_context.rs:1090]: I think it would be good to switch these assertions to use the `assert_composition_errors` helper

File: apollo-federation/src/schema/validators/from_context.rs [hunk 25]
```diff
@@ -2168,7 +2206,23 @@ mod tests {
             &context_map,
             &mut errors,
         )
-        .expect_err("unparseable fromContext directive");
+        .expect("validates fromContext directives");
+
+        // Should have error for missing context reference (no $ sign)
+        assert!(
```
Comment (tninesling) [inline apollo-federation/src/schema/validators/from_context.rs:2212]: Same note. Let's use `assert_composition_errors` with an exact string match on the message

File: apollo-federation/tests/composition/compose_set_context.rs [hunk 26]
```diff
@@ -0,0 +1,1406 @@
...
+
+    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
+
+    let errors = result.expect_err("Expected composition to fail");
+    assert_eq!(errors.len(), 1, "Expected exactly 1 error");
+
+    let error_message = errors[0].to_string();
+    assert_eq!(
+        error_message,
+        "Argument \"U.field(a:)\" is contextual in at least one subgraph but in \"U.field(a:)\" it does not have @fromContext, is not nullable and has no default value.",
```
Comment (tninesling) [inline apollo-federation/tests/composition/compose_set_context.rs:424]: This matches JS, but we should probably leave a comment and/or make a ticket that the second interpolated value probably makes more sense if it's a subgraph name

**Review by tninesling — COMMENTED** [review]
The only blocking comment is the switch from `User.id(contextArg:)` to `Entity.id`.

Also, please try to help out your reviewers by not force-pushing. When you overwrite the commits, GitHub can't track what changed since the last time we reviewed, and then we have to review the whole thing from scratch.

File: apollo-federation/src/schema/validators/from_context.rs (hunk 23 above)
Comment (briannafugate408) [reply apollo-federation/src/schema/validators/from_context.rs:1043]: We updated the message format here from just using the target to using the type name and the field, like it does in JS https://github.com/apollographql/router/pull/8504/commits/bfd0251449bea492a6a3e6f7956b220269bbb744#diff-47e4538bea230b013e832b000082609641c78eade58ecaf61ac7a25b39f3e292R660-R662

File: apollo-federation/tests/composition/compose_set_context.rs (hunk 26 above)
Comment (briannafugate408) [reply apollo-federation/tests/composition/compose_set_context.rs:424]: Added a comment and created a https://apollographql.atlassian.net/browse/FED-893

File: apollo-federation/src/schema/validators/from_context.rs (hunk 24 above)
Comment (briannafugate408) [reply apollo-federation/src/schema/validators/from_context.rs:1090]: So after trying to switch these, it seems like I wouldn't be able to update these tests to use the `assert_composition_errors` because that function works with `Result<Supergraph<Satisfiable>, Vec<CompositionError>>`, but these unit tests use MultipleFederationErrors (at least this is what the compilation error says).

File: apollo-federation/src/schema/validators/from_context.rs (hunk 23 above)
Comment (tninesling) [reply apollo-federation/src/schema/validators/from_context.rs:1043]: I see. It looks like we changed the interface name in this test compared to the JS one, but JS does output the interface field coordinate in this case instead of the implementation type 👍
